### ISSUE TITLE
Trained DaskXGBRegressor has lower accuracy than sklearn XGBRegressor

### ISSUE THREAD

**Issue by dding3** [dmlc/xgboost#5864]
I trained a sklearn.XGBRegressor with:
```
XGBRegressor(n_estimators=500, max_depth=50, n_jobs=-1, tree_method=‘hist’,
random_state=2,learning_rate=0.1, min_child_weight=1, seed=0,
subsample= 0.8, colsample_bytree=0.8, gamma=0, reg_alpha=0,
reg_lambda=1,verbosity=0)
```

I also train DaskXGBRegressor with LocalCluster(n_workers=1, threads_per_worker=1) as:
```
xgboost.dask.DaskXGBRegressor(n_estimators=500, max_depth=50, n_jobs=-1, tree_method=‘hist’,
random_state=2,learning_rate=0.1, min_child_weight=1, seed=0,
subsample= 0.8, colsample_bytree=0.8, gamma=0, reg_alpha=0,
reg_lambda=1,verbosity=0,objective=‘reg:squarederror’)
```

Data value feed to both models are the same except one is pandas dataframe while the other one is dask dataframe. And dask dataframe is converted from the pandas dataframe as:
x = dask.dataframe.from_pandas(pd_dataframe, npartitions=1)

However I find when using trained DaskXGBRegressor to do prediction, the accuracy is much lower than trained XGBRegressor.

Could this be a bug or DaskXGBRegressor was incorrectly used?

**Comment by hcho3** [comment]
Try setting subsample to 1. Distributed training could affect how the data is sampled.

**Comment by trivialfis** [comment]
Which xgboost version are you using?

**Comment by dding3** [comment]
> Which xgboost version are you using?

1.1.1, installed from pip

**Comment by trivialfis** [comment]
That's not expected.  Will verify today.

**Comment by trivialfis** [comment]
@hcho3 @RAMitchell I think we should put out a  warning about `max_depth` parameter.  The parameter used here set it to 50 and ran out of memory on my machine.

**Comment by hcho3** [comment]
@trivialfis Maybe show a warning for `max_depth > 15`?

**Comment by trivialfis** [comment]
@hcho3 Not sure, let's have discussion with some more people.  For training random forest, 15 seems to be reasonably.

**Comment by trivialfis** [comment]
Back to the topic, using default parameter set with 1 worker I was able to generate bit identical model with single node training.

**Comment by trivialfis** [comment]
It's hard to compare the result once random sampling gets in the way, as mentioned by @hcho3 the seed is set differently on distributed environment.

**Comment by dding3** [comment]
I train DaskXGBRegressor with LocalCluster(n_workers=1, threads_per_worker=1) 

> Try setting `subsample` and `colsample_bytree` to 1. Distributed training could affect how the data is sampled.

Tried setting `subsample` and `colsample_bytree` to 1. No luck :(

**Comment by dding3** [comment]
> Back to the topic, using default parameter set with 1 worker I was able to generate bit identical model with single node training.

Thank you for your verification.

I tried default parameter set(XGBRegressor() and DaskXGBRegressor()), seemed sklearn XGBRegressor still had better accuracy, MSE score: 4.1173773528105854e-05, while MSE score for DaskXGBRegressor was 0.08585785193124876. I think training set is very important here. If I only use several records as training set, the accuracy is the same.

**Comment by moritzaugustin-tiplu** [comment]
today I faced a similar issue using xgboost 1.2.1 (most recent from pypi) which might indicate that there is a serious inaccuracy issue when using more than one dask worker with xgboost (this problem currently prevents me from using xgboost in a beyond-memory dataset situation as I need accurate results): 

in my situation, using a dask `LocalCluster` with 1 worker (1 thread per worker) with xgboost leads to identical results compared with xgboost without dask which is nice and expected: the trees after the two trainings are exactly identical, and the same is true for all evaluation metrics both on test as well as evaluation set

however, when using a `LocalCluster` with 2 workers (1 thread per worker) I get different (worse) results: F1 score drops from 67% to 62% (although the dask matrices are identical  when comparing entry-by-entry between the 1 vs 2 workers scenario). also looking at the trees they deviate significantly already at the first node of the first tree a different split point was found

the xgb config dict used in all three variants (`xgb,train`, `xgb.dask.train` with 1 worker, `xgb.dask.train` with 2 workers):
`{
   'verbosity': 2,
    'tree_method': 'hist',
    'max_depth': 6,
    'nthread': 1,
}`

=> no random effects are activated, i.e., col/feature subsampling are inactive

it would be really great if an expert could look at this

I have attached a complete reproducible package: scripts, data set and usage instructions
[dask_distributed_gb.zip](https://github.com/dmlc/xgboost/files/5492414/dask_distributed_gb.zip)

**Comment by moritzaugustin** [comment]
today I faced a similar issue using xgboost 1.2.1 / dask 2.30.0 (most recent from pypi) which might indicate that there is a serious inaccuracy issue when using more than one dask worker with xgboost (this problem currently prevents me from using xgboost in a beyond-memory dataset situation as I need accurate results): 

in my situation, using a dask `LocalCluster` with 1 worker (1 thread per worker) with xgboost leads to identical results compared with xgboost without dask which is nice and expected: the trees after the two trainings are exactly identical, and the same is true for all evaluation metrics both on test as well as evaluation set

however, when using a `LocalCluster` with 2 workers (1 thread per worker) I get different (worse) results: F1 score drops from 67% to 62% (although the dask matrices are identical  when comparing entry-by-entry between the 1 vs 2 workers scenario). also looking at the trees they deviate significantly already at the first node of the first tree a different split point was found

the xgb config dict used in all three variants (`xgb,train`, `xgb.dask.train` with 1 worker, `xgb.dask.train` with 2 workers):
`{
   'verbosity': 2,
    'tree_method': 'hist',
    'max_depth': 6,
    'nthread': 1,
}`

=> no random effects are activated, i.e., col/feature subsampling are inactive

it would be really great if an expert could look at this

I have attached a complete reproducible package: scripts, data set and usage instructions:
[dask_distributed_gb.zip](https://github.com/dmlc/xgboost/files/5492467/dask_distributed_gb.zip)

**Comment by trivialfis** [comment]
![example_distributed_results_nworker2](https://user-images.githubusercontent.com/16746409/165856523-d30ace40-7c3f-4f49-a519-c082babd9ef5.png)
![example_non-distributed_results](https://user-images.githubusercontent.com/16746409/165856530-9c04f342-d42a-4594-b078-0c1c965f4e17.png)

Ran the code again, seems to be fixed now.  Note that we have removed some dependency between distributed and local training by disabling the `seed_per_iteration`. Also, for non-trivial data, the result is expected to be different due to quantile on stream algorithm.
